### DISCUSSION TITLE
Is there a java or kotlin implementation for gateway ?

### DISCUSSION THREAD

**Discussion by michaelkgeorge** [Netflix/dgs-framework#132] [Q&A]
Is there a java or kotlin implementation for gateway ?

**Comment by paulbakker** [comment]
We run our own (Kotlin) gateway implementation at Netflix, but don't currently have plans yet to open source that implementation. 
My best recommendation is to use Apollo server as the gateway - which is obviously not what you asked for :-) 

We do have an example of running DGSs behind Apollo gateway here: https://github.com/netflix/dgs-federation-example

**Reply by michaelkgeorge** [reply]
Thanks @paulbakker.  Its a shame I couldn't find any java/Kotlin implementation of gateway yet. Many frameworks in java supports federation but not gateway. Federation without gateway looks incomplete.

Will be waiting for Netflix Kotlin/Java gateway implementation :-)

**Reply by paulbakker** [reply]
I do think that the gateway can be seen as a piece of infrastructure though, just like a load balancer, a database etc. - which makes it less important in which language it's written. Your developers would still write their code in Java/Kotlin.
The more important aspect is all the tooling you get around it for schema management and such.

We built our own implementation because it needed to deeply integrate with other pieces of our infrastructure. Apollo has a great product stack that offers something similar to what we have - but available as a service/product.

**Reply by michaelkgeorge** [reply]
Makes sense. But I have a similar problem as you mentioned. The expertise, ecosystem & DevOps process all are build around java. So building a gateway layer in another language, even though its thin makes me think twice before going for it.

@paulbakker  Thank you. Very insightful and interesting conversation.

**Reply by fotoetienne** [reply]
@michaelkgeorge Thanks for expressing your interest in a JVM gateway implementation. As Paul mentioned, we don't yet have a concrete timeline for open sourcing our implementation, but knowing the level of demand from the community helps us know how to prioritize.

**Reply by RIP21** [reply]
@michaelkgeorge BTW, just my 5 cents here :) 
I do think tho that folks are right that going with TS implementation of a gateway using Apollo is the easiest way to go.
If you know Kotlin and Java, TS is mostly easy to get and many Java/Kotlin devs I know are absolutely fine with it and even love it compared to regular JS which is definitely complicated to handle after Java/Kotlin due to its dynamic nature, TS kinda helps you not to fuck up on "obvious" for a JS dev things :) 
So I'm sure you folks will be fine with TS gateway, at the end of the day it's not much of a code and I'm sure you have lot's of Frontend devs that are familiar with Node stack to some extent and will help out to set up ESLint and other stuff that will catch even more troublemaking pieces of expressions you may write being a Java dev :)
(I'm an ex Java/Groovy BigData dev converted to JS/TS stack 4.5 years ago and to FE overall, so I know the pains and feeling of mistrust you may have to JS ecosystem, it's way better than you may expect)
